### PR TITLE
[SPARK-37374][PYTHON] Fix StatCounter to use mergeStats when merging with self

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes `StatCounter` to use `mergeStats` instead of `merge` when merging with `self`.

### Why are the changes needed?

`StatCounter` should use `mergeStats` instead of `merge` when merging with `self`.

```py
>>> from pyspark.statcounter import StatCounter
>>> stats = StatCounter([1.0, 2.0, 3.0, 4.0])
>>> stats.mergeStats(stats)
Traceback (most recent call last):

...

TypeError: unsupported operand type(s) for -: 'StatCounter' and 'float'
```

This is a long standing bug but usually this bug won't be hit unless users explicitly use `mergeStats` with `self`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added some tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @dchvn @zero323 @HyukjinKwon

File: python/pyspark/tests/test_statcounter.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        self.assertDictEqual(
+            merged.asDict(),
+            {
+                "count": 8,
+                "max": 4.0,
+                "mean": 2.5,
+                "min": 1.0,
+                "stdev": 1.1952286093343936,
+                "sum": 20.0,
+                "variance": 1.4285714285714286,
```
Comment (zero323) [inline python/pyspark/tests/test_statcounter.py:68]: Just thinking out loud ‒ how robust is this? Wouldn't be better to compare individual values with `assertAlmostEqual`?

Comment (ueshin) [reply python/pyspark/tests/test_statcounter.py:68]: Sounds good!

**Review by zero323 — APPROVED** [review]
LGTM, subject to passing tests.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**